src/server/pkg/storage/chunk/reader.go

```diff
@@ -58,20 +66,27 @@ func (r *Reader) Read(data []byte) (int, error) {
 		data = data[n:]
 		totalRead += n
 		if err != nil {
-			// If all DataRefs have been read, then io.EOF.
-			if len(r.dataRefs) == 0 {
-				return totalRead, io.EOF
-			}
 			if err := r.nextDataRef(); err != nil {
 				return totalRead, err
 			}
 		}
 	}
 	return totalRead, nil
-
 }
 
 func (r *Reader) nextDataRef() error {
+	// If all DataRefs have been read, then io.EOF.
+	if len(r.dataRefs) == 0 {
+		if r.f != nil {
+			var err error
+			r.dataRefs, err = r.f()
+			if err != nil {
+				return err
+			}
```

Comment:
I think it might make sense to check here if the call to `f` returned an empty slice of dataRefs, since that shouldn't be legal, and if a callback does do that we'll segfault below when we try to access `dataRefs[0]`.